### PR TITLE
Playback controls

### PR DESCRIPTION
**Pull request by sabbaka**
It's still in progress. I plan to add current speed value somewhere and maybe somehow indicate buttons were pressed if hot key was used ( speed control ).
Space - play and pause ( button changes state ).
Arrow Right - increase playing speed.
Arrow Left - decrease playing speed.

### CODE REVIEW

**Comment by spbnick** [comment]
Thank you, Kyrill!

We need to preserve arrow left/right to control playback location in the future, when we implement the JavaScript-based player, so please don't bind them to speed decrease/increase. It would be best to have the same bindings as in tlog-play itself, when running in a terminal. If you think some of them are not good, let's discuss and change them if needed.

**Comment by sabbaka** [comment]
Ah, no, I will re-bind to { and }, I agree. I'll also add some of other commands available, basically I want to add all possible actions.

**Comment by spbnick** [comment]
Please write when you're ready for review and I'll do it.

**Comment by sabbaka** [comment]
It's ready for review. 
Please, look on behaviour of play/pause button - there is a trick to make not click-able too often ( but for now only for button, not for hot key ) and I've just realized that it might be unnecessary at all - please, check it's behaviour and let me know.
And also, take a look on play/pause button representation of playback's speed - I might move speed somewhere else, because if you'll decreasing speed, play/pause button will become wider and move button you're using to the left, so you'll click on play/pause.

**Comment by spbnick** [comment]
Sorry for the delay, I finally took a look, played with this for a while, and have a bunch of comments and requests:

* I think we should place playback controls below the terminal window, as all the players do.
* I don't think we need the stop button '⏹' - I can hardly imagine it being useful, and we have the play/pause button anyway. Please correct me if I'm wrong.
* I think we need to make restart button be "skip to start" button ('⏮'), and perhaps keep the playback control state, similar to the "skip to end" button ('⏭'). To do that I will need to implement support for specifying initial playback speed with command-line options in tlog-play. We have the rest.
* I noticed that if you press '⏭' and then '▶️', playback restarts. That shouldn't happen.
* I noticed that after playing with playback controls for a while, the host side has a bunch of tlog-play processes running. That shouldn't happen. There should be only one tlog-play process per playback window. Please investigate.
* I don't think the slow-down button should be using the "rewind" symbol ('⏪'), because it doesn't play backwards, and that is confusing. I couldn't find a standard playback control symbol for that, so even though the fast-forward button symbol ('⏩') is widely used, we might need to change it and use consistent symbols for the three speed control buttons. How about '÷2', '1:1', '×2'? Or maybe '/2', '1:1', 'x2'? In any case, we need to put those three in a group, and somewhat separately from the rest.
* We need to show tooltips with button description and hotkey, when hovering over a button.
* There are no x12 or x1/12 multipliers, as those are not powers of two. There are only x1/16, x1/8, x1/4, x1/2, x1, x2, x4, x8, x16.
* I think it is a good idea to show multiplier/divider beside the play/pause symbol. However, I think we shouldn't change the size of the button, and it might actually be better to place the divider/multiplier outside, right beside the button. Then, I think that just '/2', '/4', '/8', and '/16' are easier to read than 'x1/2', etc. Also, the the presence/absence/size of the multiplier or divider display should not affect the position of any buttons. It is hard enough to hit them with a mouse in a hurry, and even more so if they move.
* For that matter, perhaps we need to make them a bit bigger.
* Also, I think we need to place the play/pause button on the left, the other playback control buttons after a bit of space (occupied by the multiplier/divider if used), on the right, and then the speed control buttons, after a bit more space. Similar to how VLC player's UI buttons are organized.

Thank you!

**Comment by spbnick** [comment]
Also, please pass `--persist` option to tlog-play.

**Comment by sabbaka** [comment]
- Buttons moved to the bottom.
- "Stop" button removed.
- For now I've only changed icon - **still needs work.** We can keep it as is with restart of playback speed.
- Removed this behaviour from play button.
- I've not noticed this for now and might be related to play which was restarting playback -I've made some testing - have not noticed such behaviour.
- Icons changed to numbers ( /2, 1:1, x2 ).
- Added title attribute for buttons.
- Speed now is representing only power of two.
- Multiplier is now on the right side.
- Buttons are bigger now.
- I've add class to create space between buttons.

**Comment by spbnick** [comment]
@sabbaka, I just pushed the "-s/--speed" option support to tlog-play, so you can now implement rewind-to-start with fully preserved playback state. So, if rewind-to-start is pressed, whatever playback state was there should be passed to the restarted tlog-play. If it was paused, pass `--paused`, if speed was not 1:1, pass `--speed=MULTIPLIER`.

I'll take a look at your updates to this PR next.

**Comment by spbnick** [comment]
I took a look and it looks much better, thank you! I like how the buttons look and their grouping. Here are my comments and requests:

* When the play/pause button changes the picture it changes size slightly and other buttons move with it. That shouldn't happen.
* Playback doesn't react to the F key, while other keys seem to be working.
* We'll need a hotkey for rewind-to-start. Perhaps 'r' (as in "restart") or 'g' (as what "less" uses), or "B" (for "beginning", or "back" and also uppercase as in 'F' for "forward", i.e. rewind-to-end)?
* I don't think the play/pause button should be prevented from being clicked too often. What was your idea for this?

**Comment by sabbaka** [comment]
- I'll check this.
- That's strange, will check.
- Yes, I was thinking about this, I think "R" will work just fine - it's okay to have such hot-keys, like on YouTube they are using "K" for play/pause.
- I might remove this, basically it's pretty common error, it might make button state not be synced to playback, I'll test and remove if it's not the case, it's more related to interaction with back-end, which is not the case here.

**Comment by sabbaka** [comment]
I've closed all four last requests.
One more thing though - if playback is finished - play/pause button is still available to click, but nothing happens, what do you think about this? Options are: disable it or let it start playback from the beginning ( we've removed this before ).

**Comment by spbnick** [comment]
Thank you, Kyrill, this looks and feels better. Buttons staying in place gives a rock-solid feeling now.
I have some more requests before I would review the code itself, though.

* When the terminal is focused, and space is pressed, the play/pause button doesn't change state, it changes state if focus is somewhere else. Same goes for backspace. The speed display is changed when `{` or `}` is pressed, even when terminal is focused, so there is some mismatch. If it's easy, let's fix it, otherwise let's not bother.
* Please make the `⏮` button keep the playback state. I.e. keep the speed, and if playback was paused, let it stay paused. Use `--speed=NUMBER` and `--paused` tlog-play options for that. This would be useful for inspecting the very beginning of the session by stepping through/going slow, and would be less of a surprise in general.
* Please open the recording page with playback paused, i.e. start tlog-play with `--paused` option when page is opened.
* Please update tlog-play and switch to using the `G` hotkey instead of `F`. This should be the last update to tlog-play interface for now.
* Please make the restart and forward-to-end react to uppercase hotkeys only, so they match the terminal tlog-play. I.e. not just `r` and `g` keys, but `Shift-R` and `Shift-G`. For that matter, I think tooltips should display them with `Shift-`.
* Please mention the restart hotkey in the button tooltip.
* Please keep commits to a minimum of logically-separated changes. E.g. merge "Fix play button width" to the commit that introduced the button. In general, if a commit simply fixes something that was introduced by an earlier commit in the same pull request, it should be in that earlier commit.

Thank you!

**Comment by spbnick** [comment]
Regarding the play/pause button at the end, we can't detect the end of playback at the moment, so let's not bother about that. Just leave it as is. Besides, it has an effect on ongoing sessions.

**Comment by sabbaka** [comment]
In regards to play/pause and send other buttons ( but yeah, actually only play/pause matters ) - so I've tried to catch keyboard events from terminal and then send them back to the Recording component, but it did not worked - probably because of limitations of platform and window API - like when I put one more event on window for keypress - another event gone, I really want to dig into this ( but it's not so important for now ), but I am not sure whether it is possible to "catch" keypress events for div or something like this.

**Comment by sabbaka** [comment]
Also, seems that it's only related to "Space" key, I'll stash my progress and work on everything else now and then come back to this.

**Comment by sabbaka** [comment]
All requested changes are done now.

**Comment by spbnick** [comment]
Thank you, Kyrill, just a few comments:

* Even though the playback is paused when you open a recording, the play/pause button still shows up as `⏸`, as if it's playing.
* If I previously clicked the pause button, and it is still in focus, pressing space toggles it twice. We can leave this be, if there's no easy fix.
* There's still a bunch of tlog-play processes hanging around after I play with playback for a while. Perhaps abandoned on restarts? They should be cleaned up. Ping me if you're not sure why and how to deal with it.

**Comment by sabbaka** [comment]
- Oops, it's a bug.
- This can be fixed with preventDefault, will check it.
- I'll dig into this and try to reproduce on my side.

**Comment by spbnick** [comment]
Each time I hit restart a new tlog-play process is created, but the old one is not killed.

**Comment by sabbaka** [comment]
I have tried this, look what I've got https://youtu.be/DQ1KfprCjIg

**Comment by spbnick** [comment]
Do you mean your tlog-play processes don't get left behind?

**Comment by sabbaka** [comment]
`        restartPlayback() {
            if (this.state.channel != null) {
                this.state.channel.close();
            }
            this.setState({channel: this.createChannel()});
        }`


It should close process, same for componentWillUnmount and also processes will be closed if you simply close tab.

**Comment by sabbaka** [comment]
1 & 2 are resolved now. 
About tlog-play processes bug what should we do?

**Comment by spbnick** [comment]
Thank you. I can still reproduce the extra tlog-play processes. I'll investigate tomorrow morning.

**Comment by spbnick** [comment]
Regarding tlog-play processes left behind: cockpit-project/cockpit#7656

**Comment by sabbaka** [comment]
This is odd, because, it was working on my side. 
I've added "terminated" to .close() related for tlog-play in separate commit.

**Comment by spbnick** [comment]
Thank you Kyrill, I only have two last functionality comments, once we deal with them I'll review the code:

* If I first click the pause button to pause/unpause the recording, and then press space, it toggles the pause twice, essentially keeping it in the same state. From your comments I assumed you fixed it, but if I'm mistaken and you decided it's complicated, that's OK too.
* If I put playback on pause and then pressed `⏭`, it takes the pause off, but it shouldn't.

**Comment by sabbaka** [comment]
- https://youtu.be/2boVCOXB_bA - I've made test and recorded, it could cache or something ( it's possible ), so I'm pretty sure it's fixed.

**Comment by sabbaka** [comment]
I've fixed behaviour for rewind to the end.

**Comment by spbnick** [comment]
Right, regarding play/pause and space, this works fine in e.g. Google Chrome, but is still a problem in Firefox.

**Comment by spbnick** [comment]
In general, should we perhaps let go of using space to pause/play altogether? This produces too many problems and conflicts with other things. E.g. if we have focus on another button and use space to press it, it activates play/pause *and* that button and that's quite confusing.

**Comment by sabbaka** [comment]
We can use for example "P" for play / pause hot key. Space is a sort of system-wide key.
But, I can figure out why it's not working in Firefox.

**Comment by spbnick** [comment]
Perhaps we should indeed use `P` (or rather `p`, without `Shift`), and leave space working only when terminal is focused (i.e. tlog-play will be handling it). I'll make tlog-play accept `p` as well for consistency. Let's drop space for now.

**Comment by sabbaka** [comment]
I agree, because, honestly, Firefox for some reason is not reacting on preventDefault() and stopPropagation() for both .keydown and .keypress.

**Comment by sabbaka** [comment]
Done, now "p" is a hot key for play/pause.

**Comment by spbnick** [comment]
I get this from `make`:

    /home/nkondras/projects/github.com/scribery/cockpit/pkg/session_recording/recordings.jsx
      224:25  error  'event' is defined but never used  no-unused-vars

**Comment by sabbaka** [comment]
Fixed. I'm now sending space to terminal as it was, so adding p to tlog-play is only for consistency.

**Comment by spbnick** [comment]
The play/pause button tooltip still shows SPACE hotkey. Please fix.

It would be nice to send `p` to tlog-play, not space, for consistency too, but it will be fine as it is.

Otherwise this looks good, thank you :)

**Comment by sabbaka** [comment]
I should have used vars, not plain text :)
Fixed now.

**Comment by spbnick** [comment]
Great, this is fine now :) I'll review the code tomorrow morning. Thank you!

File: pkg/session_recording/recordings.css [hunk 1]
```diff
@@ -310,3 +310,11 @@ a.disabled {
 .listing > tbody > tr {
     cursor: pointer;
 }
+
+.margin-right-btn {
+    margin-right: 10px;
+}
+
+.play_btn {
```
Comment (spbnick) [inline pkg/session_recording/recordings.css]: Let's keep using dashes, not underscores, as the most of the file does.

File: pkg/session_recording/recordings.jsx [hunk 2]
```diff
@@ -161,12 +161,30 @@
...
+                playback: 'pause',
+                speed: 0,
+                wasEnded: false,
             };
         }
 
+        getSpeed(value) {
+            if (value === 0) {
+                return 1;
+            }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: 2^0 is naturally 1, so we don't need this condition here.

File: pkg/session_recording/recordings.jsx [hunk 3]
```diff
@@ -161,12 +161,30 @@
...
+            this.handleKeyDown = this.handleKeyDown.bind(this);
             this.state = {
                 channel: null,
+                playback: 'pause',
+                speed: 0,
+                wasEnded: false,
             };
         }
 
+        getSpeed(value) {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: This function is not getting anything from anywhere, but only uses its arguments, so it shouldn't be called *get*Speed, but perhaps *calc*Speed. However, given that it boils down to `Math.pow(2, value)` we might not need it at all.

File: pkg/session_recording/recordings.jsx [hunk 4]
```diff
@@ -176,14 +194,20 @@
             if (!r) {
                 return null;
             }
+            let spawn = [
+                "/usr/bin/tlog-play",
+                "--persist",
+                "--follow",
+                "--reader=journal",
+                "-M", "TLOG_REC=" + r.id,
+                "--speed=" + this.getSpeed(this.state.speed),
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: This is confusing. We already have speed in `this.state.speed`, why do we need a function to get it? For that matter, I think `this.state.speed` should be called `this.state.speed_exp` (`exp` for "exponent") instead, as that's what it is.

Comment (spbnick) [reply pkg/session_recording/recordings.jsx]: Since we don't use this function anywhere else, and given that we rename `this.state.speed` to `this.state.speed_exp`, we can just replace it with `Math.pow(2, this.state.speed_exp)`.

File: pkg/session_recording/recordings.jsx [hunk 5]
```diff
@@ -193,13 +217,77 @@
...
+
+        handleKeyDown(event) {
+            let keyCodesFuncs = {
+                "p": this.playPauseToggle,
+                "}": this.speedUp,
+                "{": this.speedDown,
+                "Backspace": this.speedReset,
+                ".": this.skipFrame,
+                "G": this.endOnce,
+                "Q": this.stopPlayback,
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: We don't need this, and the `--persist` option will prevent tlog-play from reacting to it anyway. Please remove.

File: pkg/session_recording/recordings.jsx [hunk 6]
```diff
@@ -313,16 +423,38 @@
...
+                                                <i className={"fa fa-"+ this.state.playback} aria-hidden="true"></i>
+                                                </button>
+                                            <button title="Skip Frame - Hotkey: ." type="button"
+                                                className="btn btn-default btn-lg margin-right-btn"
+                                                onClick={this.skipFrame}>
+                                                <i className="fa fa-step-forward" aria-hidden="true"></i></button>
+                                            <button title="Restart Playback - Hotkey: Shift-R" type="button"
+                                                className="btn btn-default btn-lg" onClick={this.restartPlayback}>
+                                                <i className="fa fa-fast-backward" aria-hidden="true"></i></button>
+                                            <button title="Play to the end at once - Hotkey: Shift-G" type="button"
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: I think "Fast-forward to end" would be more readable here.

File: pkg/session_recording/recordings.jsx [hunk 7]
```diff
@@ -639,7 +771,7 @@
         }
 
         render() {
-            if(this.state.error_tlog_uid === true) {
+            if (this.state.error_tlog_uid === true) {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx:752]: This should go into the "Fix "if" code style" commit.

File: pkg/session_recording/recordings.jsx [hunk 8]
```diff
@@ -193,13 +217,77 @@
             });
         }
 
+        sendToTerm(value) {
+            this.refs.terminal.state.terminal.send(value);
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Since we have our own copy of the terminal component, we might as well add a `send` function there and not have to reach into its internals here. So that this could look something like `this.refs.terminal.send(value)`.

File: pkg/session_recording/recordings.jsx [hunk 9]
```diff
@@ -161,12 +161,30 @@
...
+            this.speedDown = this.speedDown.bind(this);
+            this.speedReset = this.speedReset.bind(this);
             this.goBackToList = this.goBackToList.bind(this);
+            this.sendToTerm = this.sendToTerm.bind(this);
+            this.endOnce = this.endOnce.bind(this);
+            this.skipFrame = this.skipFrame.bind(this);
+            this.handleKeyDown = this.handleKeyDown.bind(this);
             this.state = {
                 channel: null,
+                playback: 'pause',
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Since you're not using the `playback` field for anything but pause state we can just make it a boolean and name it `paused`.

File: pkg/session_recording/recordings.jsx [hunk 10]
```diff
@@ -161,12 +161,30 @@
...
+                speed: 0,
+                wasEnded: false,
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: You don't seem to use this for anything. Can we remove it?

File: pkg/session_recording/recordings.jsx [hunk 11]
```diff
@@ -255,6 +344,27 @@
...
+                    if (this.state.speed === 0) {
+                        return '';
+                    }
+                    else if (this.state.speed > 0) {
+                        return 'x' + Math.pow(2, Math.abs(this.state.speed));
+                    }
+                    else if (this.state.speed < 0) {
+                        let speed = '' + Math.pow(2, Math.abs(this.state.speed));
+                        return '/' + speed;
+                    }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Let's make this shorter and more uniform:

    let exp = this.state.speed_exp
    let factor = Math.pow(2, Math.abs(exp));
    if (exp > 0) {
        return 'x' + factor;
    } else if (exp < 0) {
        return '/' + factor;
    } else {
        return '';
    }

File: pkg/session_recording/recordings.jsx [hunk 12]
```diff
@@ -193,13 +217,77 @@
...
+        sendToTerm(value) {
+            this.refs.terminal.state.terminal.send(value);
+        }
+
+        playPauseToggle() {
+            if (this.state.playback === 'pause') {
+                this.setState({playback: 'play'});
+            } else {
+                this.setState({playback: 'pause'});
+            }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: If we switch to the boolean `this.state.paused`, this can become just `this.setState({paused: !this.state.paused})`.

File: pkg/session_recording/recordings.jsx [hunk 13]
```diff
@@ -193,13 +217,77 @@
...
+            this.sendToTerm(' ');
+        }
+
+        speedUp() {
+            if (this.state.speed < 4) {
+                let speed = this.state.speed + 1;
+                this.setState({speed: speed});
+                this.sendToTerm('}');
+            }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: If you're going to assign speed to a variable, might as well save a little more:

    let speed = this.state.speed;
    if (speed < 4) {
        this.setState({speed: speed + 1});
        this.sendToTerm('}');
    }

File: pkg/session_recording/recordings.jsx [hunk 14]
```diff
@@ -193,13 +217,77 @@
...
+        }
+
+        speedDown() {
+            if (this.state.speed > -4) {
+                let speed = this.state.speed - 1;
+                this.setState({speed: speed});
+                this.sendToTerm('{');
+            }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Same as in `speedUp`.

File: pkg/session_recording/recordings.jsx [hunk 15]
```diff
@@ -193,13 +217,77 @@
...
+        }
+
         restartPlayback() {
             if (this.state.channel != null) {
                 this.state.channel.close();
             }
             this.setState({channel: this.createChannel()});
         }
 
+        endOnce() {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Perhaps something like `ffToEnd`, or `fastForwardToEnd` will be a better name for this function. What do you think?

File: pkg/session_recording/recordings.jsx [hunk 16]
```diff
@@ -433,7 +433,7 @@
                                             <button title="Play/Pause - Hotkey: p" type="button" ref="playbtn"
                                                 className="btn btn-default btn-lg margin-right-btn play_btn"
                                                 onClick={this.playPauseToggle}>
-                                                <i className={"fa fa-"+ this.state.playback} aria-hidden="true"></i>
+                                                <i className={"fa fa-"+ playback_state} aria-hidden="true"></i>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: This should be in the previous commit, IIUC. Anyway, if we switch to the boolean `this.state.paused`, this can become just a ternary operator: `(this.state.paused ? "play" : "pause")`.

File: pkg/session_recording/terminal.jsx [hunk 17]
```diff
@@ -127,7 +127,9 @@
         },
 
         onChannelMessage: function (event, data) {
-            this.state.terminal.write(data);
+            if(this.state.terminal) {
+                this.state.terminal.write(data);
+            }
```
Comment (spbnick) [inline pkg/session_recording/terminal.jsx]: This seems to be a candidate for a separate PR, but let's keep it here for now.

**Review by spbnick — CHANGES_REQUESTED** [review]
This looks neat, even though I have a bunch of comments/requests. Thank you!

File: pkg/session_recording/recordings.jsx (hunk 7 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx:752]: Done.

**Comment by sabbaka** [comment]
All requested changes are done now.

File: pkg/session_recording/recordings.jsx [hunk 18]
```diff
@@ -176,14 +186,20 @@
...
             }
+            let spawn = [
+                "/usr/bin/tlog-play",
+                "--persist",
+                "--follow",
+                "--reader=journal",
+                "-M", "TLOG_REC=" + r.id,
+                "--speed=" + Math.pow(2, this.state.speed_exp),
+            ];
+            if (this.state.paused === true) {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: I would say this is excessive, `if (this.state.paused) {` would be just fine :).

**Review by spbnick — CHANGES_REQUESTED** [review]
Excellent, just one comment :)

File: pkg/session_recording/recordings.jsx (hunk 18 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Agree 👍 Fixed.

**Comment by spbnick** [comment]
Merged, thank you, Kyrill!
